In [10]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile
fs = 8000

# ---------- Load signals ----------
def load_wav(path):
    fs, data = wavfile.read(path)
    data = data.astype(np.float64)
    return fs, data

fs1, m1 = load_wav("M1.wav")
fs2, m2 = load_wav("M2.wav")
fs3, m3 = load_wav("M3.wav")

In [8]:
rms1 = np.sqrt(np.mean(m1.astype(np.float64) ** 2))
rms2 = np.sqrt(np.mean(m2.astype(np.float64) ** 2))
rms3 = np.sqrt(np.mean(m3.astype(np.float64) ** 2))
print (rms1)
print (rms2)

7383.243593119038
5906.3835785917545


Part B, Q2

The sound source is closest to the microhone in the file in which the rms value is highest. In our case, rms(M1) = 7383.243593119038 and rms(M2) = 5906.3835785917545. Thus M1 is the case where the mic is closest. 

In [ ]:
def delay_correlation(d):
    start = max(0, -d)
    end = min(len(m1), len(m2) - d)
    return abs(np.sum(m1[start:end] * m2[start + d:end + d]))

max_corr = 0
best_delay = 0
for i in range(-1000, 1001):
    corr = delay_correlation(i)
    if corr > max_corr:
        max_corr = corr
        best_delay = i


print(f"The time delay then is {best_delay} samples")
print(f"Maximum correlation: {max_corr}")
print(f"Sampling frequency: {fs}")
print(f"Time delay: {best_delay/fs} seconds")

The time delay then is 4 samples
Maximum correlation: 1046598448601.0
Sampling frequency: 8000
Time delay: 0.0005 seconds
